[NameLookup] Add UnqualifiedLookupRequest

hamishknight opened the pull request:
This PR adds a request to perform unqualified lookup, and uses it to replace the `UnqualifiedLookup` type.

hamishknight committed:
Switch UnqualifiedLookup callers over to UnqualifiedLookupRequest

hamishknight committed:
Add UnqualifiedLookupRequest

This request performs raw unqualified lookup,
and will be used to replace the UnqualifiedLookup
type.

hamishknight committed:
NFC: Move LookupResult from Sema to Name Lookup

This will be used as the result type for the
unqualified lookup request.

hamishknight committed:
Remove NumUnqualifiedLookup counter

We now have an equivalent counter for
UnqualifiedLookupRequest.

hamishknight committed:
Remove UnqualifiedLookup

Now that we have UnqualifiedLookupRequest, this
class no longer serves much of a purpose. Inline
its constructor logic into the request.

hamishknight commented:
@swift-ci please test

CodaFi commented on include/swift/AST/NameLookupRequests.h:
```diff
@@ -302,6 +307,27 @@ class ExpandASTScopeRequest
+  // FIXME: We should be taking an OptionSet instead of the raw
+  // UnqualifiedLookupFlags, but we don't want to define == for OptionSet. We
+  // should find a way to define custom equality specifically for requests.
+  llvm::Expected<LookupResult> evaluate(Evaluator &evaluator, DeclName name,
```
When faced with these kinds of awful restrictions, I've found that defining a custom type helps.  You can roll back the OptionSet changes if you roll up the parameters to this request into a `LookupRequestDescriptor` or somesuch thing.

hamishknight replied:
@CodaFi Good idea! This will also allow us to regain the default parameters.

CodaFi commented on lib/Sema/TypeCheckNameLookup.cpp:
```diff
@@ -296,18 +260,18 @@ LookupResult
                                    SourceLoc loc,
                                    NameLookupOptions options) {
+  auto &ctx = dc->getASTContext();
   auto ulOptions = convertToUnqualifiedLookupOptions(options) |
```
Having this mutate the base value makes me nervous.  Can you `const`-qualify this and make copies in the blocks below?

hamishknight replied:
Given we can now pass the option set directly, I've removed the mutation.

CodaFi reviewed (approved):
Nothing but nits.  This is a fantastic foundation to build on.

hamishknight committed:
Add type to package inputs to UnqualifiedLookupRequest

This allows us use an OptionSet parameter for
the request (as currently we can't directly use it
as a parameter due to not having an == definition
for it). It also allows us to regain default
arguments for the source loc and flag parameters.

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test

hamishknight commented:
@swift-ci please smoke test compiler performance

hamishknight commented:
@swift-ci please smoke test compiler performance

CodaFi commented:
Let's take these changes and revisit the perf numbers tomorrow morning.

⛵️

CodaFi commented:
Yup, noise.
